File: backend/vyos_builders/interfaces/sstpc.py
"""
SSTP Client Interface Batch Builder

Provides all SSTPC interface batch operations.
SSTPC tunnels PPP over HTTPS to a remote SSTP server, enabling VPN
connectivity through restrictive firewalls that only allow HTTPS traffic.

No version differences exist between VyOS 1.4 and 1.5 for SSTPC.
"""

from typing import List, Dict, Any
from vyos_mappers import CommandMapperRegistry


class SstpcInterfaceBuilderMixin:
    """Complete batch builder for SSTPC interface operations."""

    _INTERNAL_BUILDER_METHODS = frozenset({
        "add_set", "add_delete", "add_multiple_sets", "clear",
        "get_operations", "operation_count", "is_empty", "get_capabilities",
    })

    def __init__(self, version: str):
        self.version = version
        self._operations: List[Dict[str, Any]] = []
        self.mappers = CommandMapperRegistry.get_all_mappers(version)
        self.interface_mapper_key = "interface_sstpc"

    # ========================================================================
    # Core Batch Operations
    # ========================================================================

    def add_set(self, path: List[str]) -> "SstpcInterfaceBuilderMixin":
        self._operations.append({"op": "set", "path": path})
        return self

    def add_delete(self, path: List[str]) -> "SstpcInterfaceBuilderMixin":
        self._operations.append({"op": "delete", "path": path})
        return self

    def add_multiple_sets(self, paths: List[List[str]]) -> "SstpcInterfaceBuilderMixin":
        for path in paths:
            self.add_set(path)
        return self

    def clear(self) -> None:
        self._operations = []

    def get_operations(self) -> List[Dict[str, Any]]:
        return self._operations.copy()

    def operation_count(self) -> int:
        return len(self._operations)

    def is_empty(self) -> bool:
        return len(self._operations) == 0

    def _mapper(self):
        return self.mappers[self.interface_mapper_key]

    # ========================================================================
    # Capabilities
    # ========================================================================

    def get_capabilities(self) -> Dict[str, Any]:
        is_v15 = "1.5" in self.version or "latest" in self.version
        return {
            "version": self.version,
            "version_info": {
                "is_1_4": not is_v15,
                "is_1_5": is_v15,
            },
            "features": {
                "description": {"supported": True, "description": "Interface description"},
                "disable": {"supported": True, "description": "Administratively disable interface"},
                "server": {"supported": True, "description": "Remote SSTP server IPv4 address or hostname"},
                "port": {"supported": True, "description": "Server port number (default: 443)"},
                "authentication": {"supported": True, "description": "Username and password for authentication"},
                "ssl_ca_certificate": {"supported": True, "description": "CA certificate from PKI for SSL verification"},
                "default_route_distance": {"supported": True, "description": "Admin distance for default route (1-255, default: 210)"},
                "no_default_route": {"supported": True, "description": "Do not install default route"},
                "no_peer_dns": {"supported": True, "description": "Do not use DNS servers provided by peer"},
                "mtu": {"supported": True, "description": "Maximum Transmission Unit in bytes (68-1500, default: 1452)"},
                "vrf": {"supported": True, "description": "VRF instance assignment"},
            },
        }

    # ========================================================================
    # Basic Interface
    # ========================================================================

    def delete_interface(self, interface: str) -> "SstpcInterfaceBuilderMixin":
        return self.add_delete(self._mapper().get_interface(interface))

    def set_interface_description(self, interface: str, description: str) -> "SstpcInterfaceBuilderMixin":
        return self.add_set(self._mapper().get_description(interface, description))

    def delete_interface_description(self, interface: str) -> "SstpcInterfaceBuilderMixin":
        return self.add_delete(self._mapper().get_description_path(interface))

    def set_interface_disable(self, interface: str) -> "SstpcInterfaceBuilderMixin":
        return self.add_set(self._mapper().get_disable(interface))

    def delete_interface_disable(self, interface: str) -> "SstpcInterfaceBuilderMixin":
        return self.add_delete(self._mapper().get_disable(interface))

    def set_vrf(self, interface: str, vrf: str) -> "SstpcInterfaceBuilderMixin":
        return self.add_set(self._mapper().get_vrf(interface, vrf))

    def delete_vrf(self, interface: str) -> "SstpcInterfaceBuilderMixin":
        return self.add_delete(self._mapper().get_vrf_path(interface))

    # ========================================================================
    # Server / Connection
    # ========================================================================

    def set_server(self, interface: str, server: str) -> "SstpcInterfaceBuilderMixin":
        return self.add_set(self._mapper().get_server(interface, server))

    def delete_server(self, interface: str) -> "SstpcInterfaceBuilderMixin":
        return self.add_delete(self._mapper().get_server_path(interface))

    def set_port(self, interface: str, port: str) -> "SstpcInterfaceBuilderMixin":
        return self.add_set(self._mapper().get_port(interface, port))

    def delete_port(self, interface: str) -> "SstpcInterfaceBuilderMixin":
        return self.add_delete(self._mapper().get_port_path(interface))

    # ========================================================================
    # Route behavior
    # ========================================================================

    def set_default_route_distance(self, interface: str, distance: str) -> "SstpcInterfaceBuilderMixin":
        return self.add_set(self._mapper().get_default_route_distance(interface, distance))

    def delete_default_route_distance(self, interface: str) -> "SstpcInterfaceBuilderMixin":
        return self.add_delete(self._mapper().get_default_route_distance_path(interface))

    def set_no_default_route(self, interface: str) -> "SstpcInterfaceBuilderMixin":
        return self.add_set(self._mapper().get_no_default_route(interface))

    def delete_no_default_route(self, interface: str) -> "SstpcInterfaceBuilderMixin":
        return self.add_delete(self._mapper().get_no_default_route(interface))

    def set_no_peer_dns(self, interface: str) -> "SstpcInterfaceBuilderMixin":
        return self.add_set(self._mapper().get_no_peer_dns(interface))

    def delete_no_peer_dns(self, interface: str) -> "SstpcInterfaceBuilderMixin":
        return self.add_delete(self._mapper().get_no_peer_dns(interface))

    # ========================================================================
    # MTU
    # ========================================================================

    def set_mtu(self, interface: str, mtu: str) -> "SstpcInterfaceBuilderMixin":
        return self.add_set(self._mapper().get_mtu(interface, mtu))

    def delete_mtu(self, interface: str) -> "SstpcInterfaceBuilderMixin":
        return self.add_delete(self._mapper().get_mtu_path(interface))

    # ========================================================================
    # Authentication
    # ========================================================================

    def set_authentication_username(self, interface: str, username: str) -> "SstpcInterfaceBuilderMixin":
        return self.add_set(self._mapper().get_authentication_username(interface, username))

    def delete_authentication_username(self, interface: str) -> "SstpcInterfaceBuilderMixin":
        return self.add_delete(self._mapper().get_authentication_username_path(interface))

    def set_authentication_password(self, interface: str, password: str) -> "SstpcInterfaceBuilderMixin":
        return self.add_set(self._mapper().get_authentication_password(interface, password))

    def delete_authentication_password(self, interface: str) -> "SstpcInterfaceBuilderMixin":
        return self.add_delete(self._mapper().get_authentication_password_path(interface))

    def delete_authentication(self, interface: str) -> "SstpcInterfaceBuilderMixin":
        return self.add_delete(self._mapper().get_authentication_path(interface))

    # ========================================================================
    # SSL
    # ========================================================================

    def set_ssl_ca_certificate(self, interface: str, ca_cert: str) -> "SstpcInterfaceBuilderMixin":
        return self.add_set(self._mapper().get_ssl_ca_certificate(interface, ca_cert))

    def delete_ssl_ca_certificate(self, interface: str) -> "SstpcInterfaceBuilderMixin":
        return self.add_delete(self._mapper().get_ssl_ca_certificate_path(interface))

    def delete_ssl(self, interface: str) -> "SstpcInterfaceBuilderMixin":
        return self.add_delete(self._mapper().get_ssl_path(interface))

File: frontend/src/lib/api/sstpc.ts
import { apiClient } from "./client";

// ============================================================================
// TypeScript Interfaces
// ============================================================================

export interface SstpcAuthentication {
  username: string | null;
  password: string | null;
}

export interface SstpcSslConfig {
  ca_certificate: string | null;
}

export interface SstpcInterface {
  name: string;
  type: string;
  description: string | null;
  disabled: boolean;
  server: string | null;
  port: string | null;
  default_route_distance: string | null;
  no_default_route: boolean;
  no_peer_dns: boolean;
  mtu: string | null;
  vrf: string | null;
  authentication: SstpcAuthentication | null;
  ssl: SstpcSslConfig | null;
}

export interface SstpcConfigResponse {
  interfaces: SstpcInterface[];
  total: number;
}

export interface SstpcCapabilityFeature {
  supported: boolean;
  description: string;
}

export interface SstpcCapabilities {
  version: string;
  version_info: {
    is_1_4: boolean;
    is_1_5: boolean;
  };
  features: Record<string, SstpcCapabilityFeature>;
}

export interface VyOSResponse {
  success: boolean;
  data?: Record<string, unknown> | null;
  error?: string | null;
}

export interface SstpcBatchOperation {
  op: string;
  value?: string;
}

export interface SstpcCreateConfig {
  name: string;
  description?: string;
  disabled?: boolean;
  server?: string;
  port?: string;
  username?: string;
  password?: string;
  ssl_ca_certificate?: string;
  default_route_distance?: string;
  no_default_route?: boolean;
  no_peer_dns?: boolean;
  mtu?: string;
  vrf?: string;
}

// ============================================================================
// Operation builder helpers
// ============================================================================

function buildInterfaceOps(config: SstpcCreateConfig): SstpcBatchOperation[] {
  const ops: SstpcBatchOperation[] = [];

  if (config.description) ops.push({ op: "set_interface_description", value: config.description });
  if (config.disabled) ops.push({ op: "set_interface_disable" });
  if (config.server) ops.push({ op: "set_server", value: config.server });
  if (config.port) ops.push({ op: "set_port", value: config.port });

  if (config.username) ops.push({ op: "set_authentication_username", value: config.username });
  if (config.password) ops.push({ op: "set_authentication_password", value: config.password });
  if (config.ssl_ca_certificate) ops.push({ op: "set_ssl_ca_certificate", value: config.ssl_ca_certificate });

  if (config.default_route_distance) ops.push({ op: "set_default_route_distance", value: config.default_route_distance });
  if (config.no_default_route) ops.push({ op: "set_no_default_route" });
  if (config.no_peer_dns) ops.push({ op: "set_no_peer_dns" });
  if (config.mtu) ops.push({ op: "set_mtu", value: config.mtu });
  if (config.vrf) ops.push({ op: "set_vrf", value: config.vrf });

  return ops;
}

// ============================================================================
// API Service
// ============================================================================

class SstpcService {
  async getCapabilities(): Promise<SstpcCapabilities> {
    return apiClient.get<SstpcCapabilities>("/vyos/sstpc/capabilities");
  }

  async getConfig(refresh: boolean = false): Promise<SstpcConfigResponse> {
    return apiClient.get<SstpcConfigResponse>("/vyos/sstpc/config", {
      refresh: refresh.toString(),
    });
  }

  async refreshConfig(): Promise<VyOSResponse> {
    return apiClient.post<VyOSResponse>("/vyos/config/refresh");
  }

  async batchConfigure(
    interfaceName: string,
    operations: SstpcBatchOperation[]
  ): Promise<VyOSResponse> {
    const result = await apiClient.post<VyOSResponse>("/vyos/sstpc/batch", {
      interface: interfaceName,
      operations,
    });
    if (result.success) {
      await this.refreshConfig();
    }
    return result;
  }

  async createInterface(config: SstpcCreateConfig): Promise<VyOSResponse> {
    const ops = buildInterfaceOps(config);
    return this.batchConfigure(config.name, ops);
  }

  async updateInterface(
    name: string,
    current: SstpcInterface,
    updated: Partial<SstpcCreateConfig>
  ): Promise<VyOSResponse> {
    const ops: SstpcBatchOperation[] = [];

    const setOrDeleteValue = (
      key: keyof Pick<
        SstpcCreateConfig,
        | "description"
        | "server"
        | "port"
        | "username"
        | "password"
        | "ssl_ca_certificate"
        | "default_route_distance"
        | "mtu"
        | "vrf"
      >,
      setOp: string,
      deleteOp: string
    ) => {
      const val = updated[key];
      if (val === undefined) return;
      if (val) ops.push({ op: setOp, value: val });
      else ops.push({ op: deleteOp });
    };

    const setOrDeleteFlag = (
      key: keyof Pick<SstpcCreateConfig, "disabled" | "no_default_route" | "no_peer_dns">,
      setOp: string,
      deleteOp: string
    ) => {
      const val = updated[key];
      if (val === undefined) return;
      ops.push({ op: val ? setOp : deleteOp });
    };

    setOrDeleteValue("description", "set_interface_description", "delete_interface_description");
    setOrDeleteFlag("disabled", "set_interface_disable", "delete_interface_disable");
    setOrDeleteValue("server", "set_server", "delete_server");
    setOrDeleteValue("port", "set_port", "delete_port");

    // Authentication: handle username and password independently
    if (updated.username !== undefined) {
      if (updated.username) ops.push({ op: "set_authentication_username", value: updated.username });
      else if (current.authentication?.username) ops.push({ op: "delete_authentication_username" });
    }
    if (updated.password !== undefined) {
      if (updated.password) ops.push({ op: "set_authentication_password", value: updated.password });
      else if (current.authentication?.password) ops.push({ op: "delete_authentication_password" });
    }

    setOrDeleteValue("ssl_ca_certificate", "set_ssl_ca_certificate", "delete_ssl_ca_certificate");
    setOrDeleteValue("default_route_distance", "set_default_route_distance", "delete_default_route_distance");
    setOrDeleteFlag("no_default_route", "set_no_default_route", "delete_no_default_route");
    setOrDeleteFlag("no_peer_dns", "set_no_peer_dns", "delete_no_peer_dns");
    setOrDeleteValue("mtu", "set_mtu", "delete_mtu");
    setOrDeleteValue("vrf", "set_vrf", "delete_vrf");

    if (ops.length === 0) {
      return { success: true, data: { message: "No changes detected" } };
    }

    return this.batchConfigure(name, ops);
  }

  async deleteInterface(name: string): Promise<VyOSResponse> {
    return this.batchConfigure(name, [{ op: "delete_interface" }]);
  }
}

export const sstpcService = new SstpcService();
